ContactList/ContactList/TableViewController.swift

```diff
@@ -33,124 +36,205 @@ class TableViewController: UITableViewController {
-        let indexPath = IndexPath(row: index, section: 0)
-        tableView.deleteRows(at: [indexPath], with: .automatic)
-        if personsArray.isEmpty {
-            tableView.deleteSections(IndexSet(arrayLiteral: 0), with: .automatic)
+    }
+    
+    private func addNewSection(with categoryName: String) {
+        if keysArray.firstIndex(of: categoryName) == nil {
+            keysArray.append(categoryName)
+            keysArray.sort()
+            groupedPersons[categoryName] = []
+            tableView.insertSections(IndexSet(arrayLiteral: getKeysArrayIndex(by: categoryName)), with: .automatic)
         }
-        tableView.endUpdates()
-        
-        WorkWithData.saveImage(by: .removed, name: id)
-        updatePersonsArray()
     }
     
-    func updatePersonsArray() {
-        WorkWithData.putArray(array: personsArray)
+    private func updateCurrentPerson(person: Person, in categoryName: String) -> Bool {
+        for dictionaryItem in groupedPersons {
+            guard let index = (dictionaryItem.value.firstIndex {(item) -> Bool in item.id == person.id}) else {
+                continue
+            }
+            let personOldVersion = groupedPersons[dictionaryItem.key]![index]
+            let sectionNumber = getKeysArrayIndex(by: dictionaryItem.key)
+            let indexPath = IndexPath(row: index, section: sectionNumber)
+            
+            if Search.isPersonsFullNameFirstCharCompare(firstPerson: personOldVersion, secondPerson: person) {
+                groupedPersons[dictionaryItem.key]![index] = person
+                tableView.reloadRows(at: [indexPath], with: .automatic)
+            } else {
+                if let targetDirectory = groupedPersons[categoryName] {
+                    let targetIndexPath = IndexPath(row: targetDirectory.count, section: getKeysArrayIndex(by: categoryName))
+                    groupedPersons[dictionaryItem.key]!.remove(at: index)
+                    groupedPersons[categoryName]?.append(person)
+                    tableView.moveRow(at: indexPath, to: targetIndexPath)
+                    tableView.reloadRows(at: [targetIndexPath], with: .automatic)
+                    if groupedPersons[dictionaryItem.key]!.isEmpty {
+                        removeSection(by: dictionaryItem.key)
+                    }
+                }
+            }
+            return true
+        }
+        return false
+    }
+    
+    private func appendNewPerson(person: Person, in categoryName: String) {
+        if let targetDirectory = groupedPersons[categoryName] {
+            let targetSectionNumber = keysArray.firstIndex(of: categoryName)!
+            let targetRow = targetDirectory.count
+            let targetIndexPath = IndexPath(row: targetRow, section: targetSectionNumber)
+            groupedPersons[categoryName]?.append(person)
+            tableView.insertRows(at: [targetIndexPath], with: .automatic)
+        }
+    }
+    
+    private func removeCurrentPerson(by id: String)    {
```

Comment:
I think if you pass `key` for section `Person` is in, it would allow you to avoid unnecessary `for` loop.